test: rewrite createreadstream tests

danieljbruce opened the pull request:
When retries are moved to google-gax from the handwritten layer of `nodejs-bigtable` then the current tests mentioned in the PR will not work. They need to be rewritten to still test retry behavior, but account for the fact that the retries will be done in a different layer.

Here is what is done to modify [`Tests in [read-rows.ts]`](https://github.com/googleapis/nodejs-bigtable/blob/17838eda19b001e322765c33a83a756eeeb68963/system-test/read-rows.ts#L126) just to clarify further:
1. This [sinon stub mock](https://github.com/googleapis/nodejs-bigtable/blob/17838eda19b001e322765c33a83a756eeeb68963/system-test/read-rows.ts#L153) currently mocks out the request function and sends responses back to createreadstream where retries are currently done. Since retries will not be done in `createreadstream` anymore and this test intends to test retries then responses must be mocked in a layer below where retries are going to be done in google-gax in a later change. The current sinon stub mock should be removed and instead the [mock grpc server](https://github.com/googleapis/nodejs-bigtable/blob/main/src/util/mock-servers/mock-server.ts) should send [responses](https://github.com/googleapis/nodejs-bigtable/blob/17838eda19b001e322765c33a83a756eeeb68963/system-test/data/read-rows-retry-test.json#L33-L36) back in order to test retry behavior.
2. As is done with the current mock, the new mock server mock should check each request to ensure it matches [request_options](https://github.com/googleapis/nodejs-bigtable/blob/17838eda19b001e322765c33a83a756eeeb68963/system-test/data/read-rows-retry-test.json#L27-L32)

danieljbruce committed:
set up the test frame

danieljbruce committed:
Test is set up to evaluate streaming behavior

danieljbruce committed:
Modify test with the mock server to pass first tst

danieljbruce committed:
Fix tests that are appending startKey and endKey

danieljbruce committed:
Getting all the tests working

- Transform the rowsLimit parameter into an integer
- Change the hook into a before hook so that we don’t attempt to create multiple mock servers
- Create a guard so that the stream only writes if there are row keys to write

danieljbruce committed:
eliminate old createreadstream test

define new interfaces too

danieljbruce committed:
Remove only. rowsLimit should be optional

danieljbruce commented on system-test/read-rows.ts:
```diff
@@ -12,69 +12,76 @@
   };
 }
 
+function getRequestOptions(request: any): google.bigtable.v2.IRowSet {
```
From the old test, this contains all the code which was used for building a request.

danieljbruce commented on system-test/read-rows.ts:
```diff
@@ -123,97 +130,89 @@ describe('Bigtable/Table', () => {
-          0,
-          'not all the responses were used'
-        );
-        assert.deepStrictEqual(requestedOptions, test.request_options);
```
These asserts are moved into `checkResults` which evaluates response/request data when the call to createreadstream results in an error or ended stream.

danieljbruce reviewed (commented):
Added a few explanations throughout the PR

danieljbruce committed:
Make rowKeysRead type more specific

danieljbruce committed:
Add after hook to shut down server

leahecole commented on system-test/read-rows.ts:
```diff
@@ -12,69 +12,76 @@
+  if (request.rows && request.rows.rowRanges) {
+    requestOptions.rowRanges = request.rows.rowRanges.map(
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      (range: any) => {
```
Is there another type option than `any` that can be used here?

danieljbruce replied:
Yup. I made this more specific and solved the resulting compiler errors too.

leahecole commented on system-test/read-rows.ts:
```diff
@@ -12,69 +12,76 @@
+    );
+  }
+  if (request.rows && request.rows.rowKeys) {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
```
Is there another type we can use besides any?

danieljbruce replied:
I got rid of this comment and made it compile. This was copy/paste from the original test along with some other any's. Now the types are more specific, but a lot of interfaces and guards needed to be added to solve compiler errors up and down the stack.

leahecole commented on system-test/read-rows.ts:
```diff
@@ -12,69 +12,76 @@
+  // shorter.
+  if (request.rowsLimit && request.rowsLimit !== '0') {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    (requestOptions as any).rowsLimit = parseInt(request.rowsLimit);
```
once again asking about the `any` 🙂

danieljbruce replied:
This one was a good one to fix. Helped catch some incorrect types downstream.

leahecole commented on system-test/read-rows.ts:
```diff
@@ -123,97 +130,93 @@ describe('Bigtable/Table', () => {
+            }
+            if (response.end_with_error) {
+              // eslint-disable-next-line @typescript-eslint/no-explicit-any
+              const error: any = new Error();
```
Is this an error coming from the server? If so is it a GoogleError not an any?

danieljbruce replied:
Yup. Changed to GoogleError.

leahecole commented on system-test/testTypes.ts:
```diff
@@ -46,3 +47,23 @@ export interface Test {
   createReadStream_options: GetRowsOptions;
 }
+
+interface CreateReadStreamResponse {
+  row_keys?: string[];
+  end_with_error?: 4;
+}
+
+interface CreateReadStreamRequest {
+  rowKeys: string[];
+  rowRanges: google.bigtable.v2.IRowRange[];
+  rowsLimit?: number;
+}
+export interface ReadRowsTest {
+  createReadStream_options?: GetRowsOptions;
+  max_retries: number;
+  responses: CreateReadStreamResponse[];
+  request_options: CreateReadStreamRequest[];
+  error: number;
+  row_keys_read: string[][];
+  name: string;
+}
```
In here there is an interesting mix of snake_case and camelCase - is there a reason why both have to be used? Or can they all be migrated to a more typical JS-y camelCase?

danieljbruce replied:
They can be migrated, but it means that the json file containing all the test data in https://github.com/googleapis/nodejs-bigtable/blob/3f5db964062502cd70235714eb7b5fb62a211d4f/system-test/data/read-rows-retry-test.json#L112 needs to be updated to match.

leahecole commented:
Is the failure of the conformance tests expected?

danieljbruce commented:
> Is the failure of the conformance tests expected?

Yup. Right now the conformance tests are supposed to be running, but a whole bunch of them fail.

danieljbruce committed:
First PR correction - remove any

danieljbruce committed:
Get rid of another any

danieljbruce committed:
Get rid of the any suppressed by typescript

It was necessary to add guards to make this work. Also changed the error type to Google error.

danieljbruce committed:
Should use CreateReadStreamRequest as the type

danieljbruce committed:
Remove the comment disabling linting on any

danieljbruce committed:
Remove the import. It is not used anymore.
